tensorboard/uploader/formatters.py

```diff
@@ -0,0 +1,100 @@
+# limitations under the License.
+# ==============================================================================
+"""Helpers that format the information about experiments as strings."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import abc
+import collections
+import json
+
+import six
+
+from tensorboard.uploader import util
+
+
+@six.add_metaclass(abc.ABCMeta)
+class BaseExperimentFormatter(object):
+    """Abstract base class for formatting experiment information as a string."""
+
+    @abc.abstractmethod
+    def format_experiment(self, experiment, experiment_url):
+        """Format the information about an experiment as a representing string.
+
+        Args:
+          experiment: An `experiment_pb2.Experiment` protobuf message for the
+            experiment to be formatted.
+          experiment_url: The URL at which the experiment can be accessed via
+            TensorBoard.
+
+        Returns:
+          A string that represents the experiment.
+        """
+        pass
+
+
+class ReadableFormatter(BaseExperimentFormatter):
+    """A formatter implementation that outputs human-readable text."""
+
+    _NAME_COLUMN_WIDTH = 12
+
+    def __init__(self):
+        super(ReadableFormatter, self).__init__()
+
+    def format_experiment(self, experiment, experiment_url):
+        output = []
+        output.append(experiment_url)
+        data = [
+            ("Name", experiment.name or "[No Name]"),
+            ("Description", experiment.description or "[No Description]"),
+            ("Id", experiment.experiment_id),
+            ("Created", util.format_time(experiment.create_time)),
+            ("Updated", util.format_time(experiment.update_time)),
+            ("Runs", str(experiment.num_runs)),
+            ("Tags", str(experiment.num_tags)),
+            ("Scalars", str(experiment.num_scalars)),
+        ]
+        for name, value in data:
+            output.append(
+                "\t%s %s" % (name.ljust(self._NAME_COLUMN_WIDTH), value,)
+            )
+        return "\n".join(output)
+
+
+class JsonFormatter(object):
+    """A formatter implementation: outputs experiment as JSON."""
+
+    _JSON_INDENT = 2
+
+    def __init__(self):
+        super(JsonFormatter, self).__init__()
+
+    def format_experiment(self, experiment, experiment_url):
+        data = [
+            ("url", experiment_url),
+            ("name", experiment.name),
+            ("description", experiment.description),
+            ("id", experiment.experiment_id),
+            ("created", util.format_time(experiment.create_time)),
+            ("updated", util.format_time(experiment.update_time)),
+            ("runs", experiment.num_runs),
+            ("tags", experiment.num_tags),
+            ("scalars", experiment.num_scalars),
+        ]
+        return json.dumps(
+            collections.OrderedDict(data), indent=self._JSON_INDENT,
```

Comment:
Optional, but: if you’re concerned about ease of parsing the JSON output
without `jq`, you could consider using `indent=None` (the default), so
that the output is one JSON object per line. This makes it easy to parse
with a parser that can only take a complete JSON string, like Python’s
`json.loads` or JS’s `JSON.parse`, because the user can easily identify
the framing boundaries (just split by newline). Otherwise, they have to
identify the actual object boundaries themselves, which is slightly less
trivial.

Up to you; just mentioning because you noted this concern in the
original PR description.

Reply:
I did a test with `tensorboard dev list --json | jq -s` under this PR. `jq` seems to be able to handle the `indent=2` just fine. So I'll leave it as is.

Reply:
Right, yes, `jq` handles it fine, hence “[…] ease of parsing the JSON
output without `jq`.” This is fine with me.

Reply:
Ack. I think the slight difficulty of parsing with other tools should be acceptable.